File: server/led/morse.py
# International Morse code + standard timing.
#
# Timing is measured in "units" (a dot = 1 unit):
#   dot                 = 1 unit ON
#   dash                = 3 units ON
#   gap between symbols = 1 unit OFF   (within a letter)
#   gap between letters = 3 units OFF
#   gap between words   = 7 units OFF
#
# "Standard speed" uses the PARIS convention: unit_ms = 1200 / WPM.

MORSE = {
    "A": ".-", "B": "-...", "C": "-.-.", "D": "-..", "E": ".",
    "F": "..-.", "G": "--.", "H": "....", "I": "..", "J": ".---",
    "K": "-.-", "L": ".-..", "M": "--", "N": "-.", "O": "---",
    "P": ".--.", "Q": "--.-", "R": ".-.", "S": "...", "T": "-",
    "U": "..-", "V": "...-", "W": ".--", "X": "-..-", "Y": "-.--",
    "Z": "--..",
    "0": "-----", "1": ".----", "2": "..---", "3": "...--", "4": "....-",
    "5": ".....", "6": "-....", "7": "--...", "8": "---..", "9": "----.",
    ".": ".-.-.-", ",": "--..--", "?": "..--..", "'": ".----.",
    "!": "-.-.--", "/": "-..-.", "(": "-.--.", ")": "-.--.-",
    "&": ".-...", ":": "---...", ";": "-.-.-.", "=": "-...-",
    "+": ".-.-.", "-": "-....-", "_": "..--.-", '"': ".-..-.",
    "@": ".--.-.",
}


def units_per_wpm(wpm):
    """Milliseconds per Morse unit at the given words-per-minute."""
    wpm = max(1, min(60, int(wpm)))
    return 1200 // wpm


def build_timeline(message, unit_ms):
    """
    Turn a message into a list of (level, end_ms) segments plus the total
    cycle length. `level` is 1 (LED on) or 0 (off); `end_ms` is the cumulative
    end time of that segment. A trailing word gap separates repeats.
    Returns ([], 0) when nothing is renderable.
    """
    events = []  # (level, duration_ms)
    first_letter = True

    for ch in message.upper():
        if ch == " ":
            events.append((0, 7 * unit_ms))
            first_letter = True
            continue

        code = MORSE.get(ch)
        if not code:
            continue

        if not first_letter:
            events.append((0, 3 * unit_ms))  # inter-letter gap
        first_letter = False

        for i, sym in enumerate(code):
            events.append((1, (3 if sym == "-" else 1) * unit_ms))
            if i < len(code) - 1:
                events.append((0, 1 * unit_ms))  # intra-character gap

    if not events:
        return [], 0

    events.append((0, 7 * unit_ms))  # gap before the message repeats

    segments = []
    total = 0
    for level, dur in events:
        if dur <= 0:
            continue
        total += dur
        segments.append((level, total))
    return segments, total

File: web/src/lib/morse.js
// Mirrors server/led/morse.py so the client preview matches the real LED.

export const MORSE = {
  A: '.-', B: '-...', C: '-.-.', D: '-..', E: '.', F: '..-.', G: '--.', H: '....',
  I: '..', J: '.---', K: '-.-', L: '.-..', M: '--', N: '-.', O: '---', P: '.--.',
  Q: '--.-', R: '.-.', S: '...', T: '-', U: '..-', V: '...-', W: '.--', X: '-..-',
  Y: '-.--', Z: '--..',
  0: '-----', 1: '.----', 2: '..---', 3: '...--', 4: '....-',
  5: '.....', 6: '-....', 7: '--...', 8: '---..', 9: '----.',
  '.': '.-.-.-', ',': '--..--', '?': '..--..', "'": '.----.', '!': '-.-.--',
  '/': '-..-.', '(': '-.--.', ')': '-.--.-', '&': '.-...', ':': '---...',
  ';': '-.-.-.', '=': '-...-', '+': '.-.-.', '-': '-....-', _: '..--.-',
  '"': '.-..-.', '@': '.--.-.',
}

// Milliseconds per Morse unit at the given WPM (PARIS standard), clamped like
// the firmware.
export function unitMs(wpm) {
  const w = Math.max(1, Math.min(60, Math.round(wpm)))
  return Math.max(1, Math.floor(1200 / w))
}

// Human-readable dot/dash string (space -> "/").
export function morseString(message) {
  return message
    .toUpperCase()
    .split('')
    .map((c) => (c === ' ' ? '/' : MORSE[c] || ''))
    .filter(Boolean)
    .join(' ')
}

// Ordered [{level, dur}] segments that loop, matching build_timeline() on the
// device: dot=1u, dash=3u, symbol gap=1u, letter gap=3u, word gap=7u, plus a
// trailing word gap before the message repeats.
export function buildSegments(message, unit) {
  const events = []
  let firstLetter = true

  for (const ch of message.toUpperCase()) {
    if (ch === ' ') {
      events.push({ level: 0, dur: 7 * unit })
      firstLetter = true
      continue
    }
    const code = MORSE[ch]
    if (!code) continue

    if (!firstLetter) events.push({ level: 0, dur: 3 * unit })
    firstLetter = false

    for (let i = 0; i < code.length; i++) {
      events.push({ level: 1, dur: (code[i] === '-' ? 3 : 1) * unit })
      if (i < code.length - 1) events.push({ level: 0, dur: unit })
    }
  }

  if (!events.length) return []
  events.push({ level: 0, dur: 7 * unit })
  return events
}
